---
title: >
  Setting up plateletopedia as a resource
author:
- name: Federico Marini
  affiliation: 
  - Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Annekathrin Ludt
  affiliation: 
  - Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  email: anneludt@uni-mainz.de
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('plateletopedia')`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Setting up plateletopedia as a resource}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{plateletopedia}
  %\VignetteKeywords{GeneExpression, RNASeq, FunctionalAnnotation, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

<style type="text/css">
.smaller {
  font-size: 10px
}
</style>

**Compiled date**: `r Sys.Date()`

**Last edited**: 2021-07-14

**License**: `r packageDescription("plateletopedia")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment  = "#>",
  error    = FALSE,
  warning  = FALSE,
  eval     = FALSE,
  message  = FALSE,
  fig.width = 10
)
options(width = 100)
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
```

---

# Setup {#setup}

## Retrieval of references

```{r}
setup_folders(project_name = "plateletopedia")

retrieve_ensembl_refs(version_number = 102, species = "human", just_check = FALSE)
retrieve_ensembl_refs(version_number = 102, species = "mouse", just_check = FALSE)

retrieve_ensembl_refs(version_number = 104, species = "human", just_check = FALSE)
retrieve_ensembl_refs(version_number = 104, species = "mouse", just_check = FALSE)



retrieve_gencode_refs(version_number = "38",just_check = FALSE)
retrieve_gencode_refs(version_number = "M27",just_check = FALSE)

retrieve_gencode_refs(version_number = "37",just_check = FALSE)
retrieve_gencode_refs(version_number = "M25",just_check = FALSE)


```

## Creating indices for alignment and quantifications

```{bash}
salmon --version

## for Ensembl

grep "^>" <(gunzip -c plateletopedia_ref/Ensembl.GRCh38.102/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/Ensembl.GRCh38.102/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/Ensembl.GRCh38.102/decoys.txt

grep "^>" <(gunzip -c plateletopedia_ref/Ensembl.GRCh38.104/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/Ensembl.GRCh38.104/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/Ensembl.GRCh38.104/decoys.txt


grep "^>" <(gunzip -c plateletopedia_ref/Ensembl.GRCm38.102/Mus_musculus.GRCm38.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/Ensembl.GRCm38.102/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/Ensembl.GRCm38.102/decoys.txt

grep "^>" <(gunzip -c plateletopedia_ref/Ensembl.GRCm39.104/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/Ensembl.GRCm39.104/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/Ensembl.GRCm39.104/decoys.txt



## for Gencode

grep "^>" <(gunzip -c plateletopedia_ref/GENCODE_37/GRCh38.primary_assembly.genome.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/GENCODE_37/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/GENCODE_37/decoys.txt

grep "^>" <(gunzip -c plateletopedia_ref/GENCODE_38/GRCh38.primary_assembly.genome.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/GENCODE_38/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/GENCODE_38/decoys.txt


grep "^>" <(gunzip -c plateletopedia_ref/GENCODE_M25/GRCm38.primary_assembly.genome.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/GENCODE_M25/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/GENCODE_M25/decoys.txt

grep "^>" <(gunzip -c plateletopedia_ref/GENCODE_M27/GRCm39.primary_assembly.genome.fa.gz) | cut -d " " -f 1 > plateletopedia_ref/GENCODE_M27/decoys.txt
sed -i.bak -e 's/>//g' plateletopedia_ref/GENCODE_M27/decoys.txt





cat plateletopedia_ref/GENCODE_37/gencode.v37.transcripts.fa.gz plateletopedia_ref/GENCODE_37/GRCh38.primary_assembly.genome.fa.gz > plateletopedia_ref/GENCODE_37/gentrome.fa.gz
cat plateletopedia_ref/GENCODE_38/gencode.v38.transcripts.fa.gz plateletopedia_ref/GENCODE_38/GRCh38.primary_assembly.genome.fa.gz > plateletopedia_ref/GENCODE_38/gentrome.fa.gz
cat plateletopedia_ref/GENCODE_M25/gencode.vM25.transcripts.fa.gz plateletopedia_ref/GENCODE_M25/GRCm38.primary_assembly.genome.fa.gz > plateletopedia_ref/GENCODE_M25/gentrome.fa.gz
cat plateletopedia_ref/GENCODE_M27/gencode.vM27.transcripts.fa.gz plateletopedia_ref/GENCODE_M27/GRCm39.primary_assembly.genome.fa.gz > plateletopedia_ref/GENCODE_M27/gentrome.fa.gz



salmon index -t plateletopedia_ref/GENCODE_37/gentrome.fa.gz -d plateletopedia_ref/GENCODE_37/decoys.txt -p 12 -i plateletopedia_ref/salmon_index_gentrome_GENCODE_37 --gencode
salmon index -t plateletopedia_ref/GENCODE_38/gentrome.fa.gz -d plateletopedia_ref/GENCODE_38/decoys.txt -p 12 -i plateletopedia_ref/salmon_index_gentrome_GENCODE_38 --gencode
salmon index -t plateletopedia_ref/GENCODE_M25/gentrome.fa.gz -d plateletopedia_ref/GENCODE_M25/decoys.txt -p 12 -i plateletopedia_ref/salmon_index_gentrome_GENCODE_M25 --gencode
salmon index -t plateletopedia_ref/GENCODE_M27/gentrome.fa.gz -d plateletopedia_ref/GENCODE_M27/decoys.txt -p 12 -i plateletopedia_ref/salmon_index_gentrome_GENCODE_M27 --gencode
```

